### PR TITLE
Fixes Bug in cpp extension test with complex float tensor

### PR DESCRIPTION
**Pull request by soumyarooproy**
* Modify the testcase as outlined in the issue
   * Issue url: https://github.com/pytorch/pytorch/issues/14829

### CODE REVIEW

**Comment by soumyarooproy** [comment]
@goldsborough for review.

File: test/cpp_extensions/complex_registration_extension.cpp [hunk 1]
```diff
@@ -40,21 +40,9 @@ struct CPUComplexFloatType : public at::CPUTypeDefault {
...
-        scalarTypeToTypeMeta(ScalarType::ComplexFloat),
-        numel,
-        getCPUAllocator(),
-        /* resizable */ true)};
-    Tensor t{c10::make_intrusive<TensorImpl, UndefinedTensorImpl>(
-        std::move(s),
-        at::CPUTensorId(),
-        /* is_variable */ false)};
-    return t;
+    // Delegate to the appropriate cpu tensor factory
```
Comment (goldsborough) [inline test/cpp_extensions/complex_registration_extension.cpp]: Aren't you getting rid of all the code that makes the resulting tensor be of `complex` type?

Comment (soumyarooproy) [reply test/cpp_extensions/complex_registration_extension.cpp]: Correct me if I am wrong but my understanding is that the `dtype=torch.complex64` at the python call-site (as in the test) is captured in the `TensorOptions` object argument that's passed to this `empty()` factory. The `empty_cpu()` function extracts the appropriate scalar type from the options. Refer to this:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorFactories.cpp#L113

File: test/test_cpp_extensions.py [hunk 2]
```diff
@@ -283,7 +283,9 @@ def test_complex_registration(self):
             sources='cpp_extensions/complex_registration_extension.cpp',
             verbose=True)
 
-        torch.empty(2, 2, dtype=torch.complex64)
+        # Make sure that the empty tensor is of the desired shape
+        # Refer to https://github.com/pytorch/pytorch/issues/14829
+        self.assertEqual(torch.empty(2, 2, dtype=torch.complex64).size(), torch.Size([2, 2]))
```
Comment (soumyarooproy) [inline test/test_cpp_extensions.py]: Planning to change this to:
```
        t = torch.empty(2, 2, dtype=torch.complex64)
        self.assertEqual(t.size(), torch.Size([2, 2]))
        self.assertEqual(t.type(), 'torch.ComplexFloatTensor')
```
What do you think?

**Comment by soumyarooproy** [comment]
@goldsborough for a another look

**Comment by goldsborough** [comment]
@ezyang could you take a quick look? You wrote the original complex type test.

**Review by ezyang — APPROVED** [review]

**Comment by soumyarooproy** [comment]
@goldsborough @ezyang @soumith — IIANM, the ball's now in your court to merge and close this PR.

**Pull request closed without merging**